File: 01.DFS递归/leetcode_3286_findSafeWalk.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <functional>
using namespace std;

class Solution {
public:
    bool findSafeWalk(vector<vector<int>>& grid, int health) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> visited(m, vector<int>(n, 0));
        unordered_map<long long, bool> mp;
        auto dfs = [&](auto&& dfs, int i, int j){
            health -= grid[i][j];
            if (health <= 0){
                health += grid[i][j];
                return false;
            }
            long long key = (long long)i<<32 | (long long)j<<16 | health;
            if (mp.count(key)){
                health += grid[i][j];
                return mp[key];
            }
            visited[i][j] = 1;
            if (i == m-1 && j == n-1){
                return true;
            }
            bool res = false;
            if (i > 0 && visited[i-1][j] == 0){
                res |= dfs(dfs, i-1, j);
            }
            if (i < m-1 && visited[i+1][j] == 0){
                res |= dfs(dfs, i+1, j);
            }
            if (j > 0 && visited[i][j-1] == 0){
                res |= dfs(dfs, i, j-1);
            }
            if (j < n-1 && visited[i][j+1] == 0){
                res |= dfs(dfs, i, j+1);
            }
            visited[i][j] = 0;
            health += grid[i][j];
            mp[key] = res;
            return res;
        };
        return dfs(dfs, 0, 0);
    }
};

int main(){
    Solution s;
    vector<vector<int>> grid = {{0,1,0,0,0},{0,1,0,1,0},{0,0,0,1,0}};
    int health = 1;
    cout << s.findSafeWalk(grid, health) << endl;
}

File: 16.Dijkstra最短路/leetcode_3286_findSafeWalk.py
from typing import List
from heapq import *

class Solution:
    def findSafeWalk(self, grid: List[List[int]], health: int) -> bool:
        # 找到最小需要的健康值，和health比较大小即可
        m, n = len(grid), len(grid[0])
        q = [(grid[0][0],0,0)]
        dist = [[10**18]*n for _ in range(m)]
        dist[0][0] = grid[0][0]
        dirs = [(0,1), (0,-1), (1,0), (-1,0)]
        visited = [[0]*n for _ in range(m)]
        while q:
            d, x, y = heappop(q)
            if visited[x][y]:
                continue
            visited[x][y] = 1
            for dx,dy in dirs:
                nx, ny = x+dx, y+dy
                if nx<0 or nx>=m or ny<0 or ny>=n:
                    continue
                curD = d+grid[nx][ny]
                if curD < dist[nx][ny]:
                    dist[nx][ny] = curD
                    heappush(q, (curD, nx, ny))
        
        return dist[m-1][n-1] < health

if __name__ == "__main__":
    s = Solution()
    grid = [[0,1,0,0,0],[0,1,0,1,0],[0,0,0,1,0]]
    health = 1
    print(s.findSafeWalk(grid, health))
